subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultConfigurationResolver.java

```diff
@@ -191,6 +202,82 @@ public void resolveArtifacts(ConfigurationInternal configuration, ResolverResult
         results.artifactsResolved(new DefaultResolvedConfiguration(result), result);
     }
 
+    @Override
+    public List<ArtifactRepository> getRepositories() {
+        List<ExposableRepository> exposableRepositories = CollectionUtils.collect(repositories, Transformers.cast(ExposableRepository.class));
+        List<ArtifactRepository> result = Lists.newArrayListWithExpectedSize(exposableRepositories.size());
+        for (ExposableRepository repository : exposableRepositories) {
+            ConfiguredModuleComponentRepository resolver = repository.createResolver();
+            result.add(ArtifactRepositoryImpl.from(
+                resolver.getId(),
+                getTypeOf(repository),
+                resolver.getName(),
+                repository.getProperties()
+            ));
+        }
+        return result;
+    }
+
+    private static class ArtifactRepositoryImpl implements ArtifactRepository {
+
+        private final String repositoryId;
+        private final String type;
+        private final String name;
+        private final Map<String, ?> properties;
+
+        private ArtifactRepositoryImpl(String repositoryId, RepositoryType type, String name, Map<String, ?> properties) {
+            this.repositoryId = repositoryId;
+            this.type = type.displayName;
+            this.name = name;
+            this.properties = ImmutableMap.copyOf(properties);
+        }
+
+        private static ArtifactRepositoryImpl from(String repositoryId, RepositoryType type, String name, Map<RepositoryPropertyType, ?> properties) {
+            Map<String, Object> props = new HashMap<String, Object>(properties.size());
+            for (Map.Entry<RepositoryPropertyType, ?> entry : properties.entrySet()) {
+                props.put(entry.getKey().displayName, entry.getValue());
+            }
+            return new ArtifactRepositoryImpl(
+                repositoryId,
+                type,
+                name,
+                props
+            );
+        }
+
+        @Override
+        public String getRepositoryId() {
+            return repositoryId;
+        }
+
+        @Override
+        public String getType() {
+            return type;
+        }
+
+        @Override
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public Map<String, ?> getProperties() {
+            return properties;
+        }
+    }
+
+    private RepositoryType getTypeOf(ExposableRepository repository) {
+        if (repository instanceof IvyArtifactRepository) {
+            return RepositoryType.IVY;
+        } else if (repository instanceof MavenArtifactRepository) {
+            return RepositoryType.MAVEN;
+        } else if (repository instanceof FlatDirectoryArtifactRepository) {
+            return RepositoryType.FLAT_DIR;
+        } else {
+            throw new IllegalArgumentException("unhandled repository type: " + repository.getClass());
```

Comment:
Actually we should probably add smoke tests with some external plugins that could implement their own repository types. Ask @blindpirate if he had to implement one. I think the Clojure plugin does something with repositories too, and potentially a Python one I don't recall. I'm not sure they actually create a custom type, but if they do then we shouldn't fail here, but rather return something like "Unsupported".

Reply:
mmm good point. I'm not used to this possibility of leakage. Even without smoke testing, I think we should do the 'unsupported' thing you propose.